### PR TITLE
New Framework

### PR DESCRIPTION
**Pull request by shahbuland**
What I've done:
- Implementation of ILQL on IMDB positive review generation task (it runs and loss goes down, but haven't fully trained yet)
- Overall framework and utils that should be extendable to other algorithms
- Configs for training/models
- Example scripts to make use of models and show how certain utilities should be used and how training should be done

To still be done:
- PPO implementation
- Accelerate/multigpu support (Next priority) + mixed precision
- Full training of ILQL on sentiment task to test it out  

Let me know if you believe this is good to merge, or if I should finish some of the above features first.

### CODE REVIEW

File: framework/configs.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    model_path : str
+    model_arch : str # One of the architectures present in framework.model
+    device : str
+    
+    @classmethod
+    def from_dict(cls, config: Dict[str, Any]):
+        return cls(**config)
+
+@dataclass
+class TrainConfig:
```
Comment (LouisCastricato) [inline framework/configs.py:16]: These should have default values!

File: framework/configs.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from dataclasses import dataclass
+from typing import Any, Dict
+import yaml
+
+@dataclass
+class ModelConfig:
```
Comment (LouisCastricato) [inline framework/configs.py:6]: Do we want a model type parameter here?

File: framework/data/__init__.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    Batch element for Gyarados or Gyarados-like similarity scoring model
+    """
+    content : Any = None
+    preference : Any = None
+    score : float = None
+
+@dataclass
+class RLElement:
+    state : Any = None
```
Comment (LouisCastricato) [inline framework/data/__init__.py:23]: State and action should not be any

File: framework/data/__init__.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    action : Any = None
+    reward : float = None
+
+@dataclass
+class BatchElement:
+    """
+    General batch element for any transformer to use in its forward pass
+    """
+    tokens : TensorType["BATCH", "SEQ_LEN"]
+    masks : TensorType["BATCH", "SEQ_LEN"]
```
Comment (LouisCastricato) [inline framework/data/__init__.py:33]: please always end files in a new line

File: framework/data/sentiment.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from dataclasses import dataclass
+
+from typing import Iterable, Callable
+from torchtyping import TensorType
+
+import torch
+
+@dataclass
+class SentimentGeneralElement:
```
Comment (LouisCastricato) [inline framework/data/sentiment.py:9]: Comments!

File: framework/data/text_gen.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+from typing import List, Callable, Iterable
+
+from framework.data import SimElement, RLElement, GeneralElement
+
+@dataclass
+class TextGeneralElement(GeneralElement):
+    text : str = None
+
+@dataclass
+class TextSimBatchElement(SimElement):
```
Comment (LouisCastricato) [inline framework/data/text_gen.py:12]: More comments

File: framework/data/text_gen.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    content : Iterable[str] = None
+    preference : Iterable[str] = None
+
+@dataclass
+class TextRLElement(RLElement):
+    state : str = None
+    action : int = None
+    reward : float = None
```
Comment (LouisCastricato) [inline framework/data/text_gen.py:20]: end in new line

File: framework/eval/sentiment.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    sentiment_chunk_size = min(512, total_samples)
+
+    with torch.no_grad():
+        samples = [model.sample(length = 32, n_samples = sample_chunk_size) for _ in range(total_samples//sample_chunk_size)]
+        samples = flatten(samples)
+        samples_chunks = chunk(samples, sentiment_chunk_size)
+
+        pipe_device = 0 if torch.cuda.is_available() else -1
+        pipe = PIPE('sentiment-analysis', 'lvwerra/distilbert-imdb', device=pipe_device)
+        sentiments = [pipe(chunk) for chunk in samples_chunks]
```
Comment (LouisCastricato) [inline framework/eval/sentiment.py:28]: comments

File: framework/eval/sentiment.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+from transformers import pipeline as PIPE
```
Comment (LouisCastricato) [inline framework/eval/sentiment.py:1]: Why is pipe caps!

File: framework/orchestrator/sentiment.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+@register_orchestrator
+class OfflineSentimentOrchestrator(Orchestrator):
+    def __init__(self, pipeline : SentimentPipeline, rl_model : BaseRLModel):
+        self.pipeline = pipeline
+        self.rl_model = rl_model
+
+        pipe_device = 0 if torch.cuda.is_available() else -1
+        self.sentiment_pipe = PIPE('sentiment-analysis', 'lvwerra/distilbert-imdb', device=pipe_device)
+
+    def make_experience(self, chunk_size = 512):
```
Comment (LouisCastricato) [inline framework/orchestrator/sentiment.py:20]: Comments!!!

**Review by LouisCastricato — CHANGES_REQUESTED** [review]

File: framework/configs.py (hunk 2 above)
Comment (shahbuland) [reply framework/configs.py:6]: model_arch is model_type 
arch as in architecture, though maybe changing to model_type would be more clear. done that

File: framework/eval/sentiment.py (hunk 9 above)
Comment (shahbuland) [reply framework/eval/sentiment.py:1]: Was a bandaid fix to resolve issue of pipeline being defined in two places once. Will change to more meaningful name.

File: framework/data/sentiment.py (hunk 5 above)
Comment (shahbuland) [reply framework/data/sentiment.py:9]: Should I add comments to classes where their superclass has all the nessecary documentation already? It doesn't seem particularly helpful in that circumstance, though I've gone ahead and added comments where they seemed appropriate.

File: framework/eval/sentiment.py (hunk 8 above)
Comment (shahbuland) [reply framework/eval/sentiment.py:32]: Have added (for this and orchestrator), check to see if they're sufficient.

Comment (LouisCastricato) [reply framework/eval/sentiment.py:32]: lgtm

**Comment by LouisCastricato** [comment]
Why did you rename PIPE to tfpipe? what does tf stand for?

**Comment by shahbuland** [comment]
> Why did you rename PIPE to tfpipe? what does tf stand for?

transformers.pipeline => transformerspipeline => tfpipeline? 😅
I was trying to avoid renaming the pipeline parameter to orchestrator.
